tests/array_tests.cpp

```diff
@@ -69,6 +69,25 @@ TEST_CASE("test array basics") {
   CHECK(x.shape() == std::vector<int>{1, 3});
   CHECK(x.strides() == std::vector<size_t>{3, 1});
 
+  int dim_min = std::numeric_limits<int>::min();
+  int dim_max = std::numeric_limits<int>::max();
+  CHECK_EQ(array::check_shape_dim(-4), -4);
+  CHECK_EQ(array::check_shape_dim(0), 0);
+  CHECK_EQ(array::check_shape_dim(12), 12);
+  CHECK_EQ(array::check_shape_dim(static_cast<ssize_t>(dim_min)), dim_min);
+  CHECK_EQ(array::check_shape_dim(static_cast<ssize_t>(dim_max)), dim_max);
+  CHECK_EQ(array::check_shape_dim(static_cast<size_t>(0)), 0);
+  CHECK_EQ(array::check_shape_dim(static_cast<size_t>(dim_max)), dim_max);
+  CHECK_THROWS_AS(
+      array::check_shape_dim(static_cast<ssize_t>(dim_min) - 1),
+      std::invalid_argument);
+  CHECK_THROWS_AS(
+      array::check_shape_dim(static_cast<ssize_t>(dim_max) + 1),
+      std::invalid_argument);
+  CHECK_THROWS_AS(
+      array::check_shape_dim(static_cast<size_t>(dim_max) + 1),
+      std::invalid_argument);
```

Comment:
Maybe make this a python test and more integrated, something that tests e.g.:

```
>>> a = np.empty(2**32, np.bool_)
>>> mx.array(a).shape
(0,)
```

Reply:
That's only 4GB so it's not too bad I think..

Reply:
I've been burned by large allocations before, so excluded such tests. If we move `check_shape_dim` to `python/src/array.cpp`, will try adding python tests.

Reply:
Added python test case.